### PR TITLE
Enable Pitest plugin for ServiceTalk

### PR DESCRIPTION
**Pull request by bondolo**
Motivation:
[PIT](https://pitest.org) provides mutation testing which is a good measure of the completeness of a project's API tests. As ServiceTalk has a large and complex API it is worth testing to ensure that the tests are complete and adequate. The quality of the API and tests can be improved by testing currently unhandled cases.

Modifications:
A [PIT gradle plugin](https://gradle-pitest-plugin.solidsoft.info) which provides PIT integration is added to the build. Because PIT testing is extremely slow the plugin is not currently added to the quality task.

Result:
The `pitest` task is now available. The reporting by the PITest plugin is currently limited but is available in the module
directory `build/pitest/index.html` after executing the `pitest` task.

### CODE REVIEW

File: servicetalk-gradle-plugin-internal/src/main/groovy/io/servicetalk/gradle/plugin/internal/ProjectUtils.groovy [hunk 1]
```diff
@@ -211,6 +211,9 @@ final class ProjectUtils {
         if (tasks.findByName("checkstyle")) {
           dependsOn tasks.checkstyle
         }
+        if (tasks.findByName("pitest")) {
+          dependsOn tasks.pitest
```
Comment (idelpivnitskiy) [inline servicetalk-gradle-plugin-internal/src/main/groovy/io/servicetalk/gradle/plugin/internal/ProjectUtils.groovy:215]: > Because PIT testing is extremely slow the plugin is not currently added to the quality task.

It is added. Let's remove this dependency for now.

File: gradle.properties [hunk 2]
```diff
@@ -46,6 +46,7 @@ protobufVersion=3.15.3
 protoGoogleCommonProtosVersion=2.0.1
 javaPoetVersion=1.13.0
 shadowPluginVersion=4.0.4
+pitestPluginVersion=1.6.0
```
Comment (idelpivnitskiy) [inline gradle.properties:49]: This is a test dependency, consider moving it under the `Test dependencies` section  , before/after `spotbugsPluginVersion`.

File: servicetalk-gradle-plugin-internal/src/main/groovy/io/servicetalk/gradle/plugin/internal/ServiceTalkLibraryPlugin.groovy [hunk 3]
```diff
@@ -206,6 +207,18 @@ final class ServiceTalkLibraryPlugin extends ServiceTalkCorePlugin {
     }
   }
 
+  private static void applyPitestPlugin(Project project) {
+    project.configure(project) {
+      pluginManager.apply("info.solidsoft.pitest")
+
+      pitest {
+        timestampedReports = false
+        outputFormats = ['XML', 'HTML']
```
Comment (idelpivnitskiy) [inline servicetalk-gradle-plugin-internal/src/main/groovy/io/servicetalk/gradle/plugin/internal/ServiceTalkLibraryPlugin.groovy:216]: spotbugs uses xml for CI and html for non-CI builds, see `applySpotBugsPlugin`. there is no need to have both

File: servicetalk-gradle-plugin-internal/src/main/groovy/io/servicetalk/gradle/plugin/internal/Versions.groovy [hunk 4]
```diff
@@ -21,6 +21,7 @@ import static org.gradle.api.JavaVersion.VERSION_1_8
 
 final class Versions {
   static final String CHECKSTYLE_VERSION = "8.41.1"
+  static final String PITEST_VERSION = "1.6.0"
```
Comment (idelpivnitskiy) [inline servicetalk-gradle-plugin-internal/src/main/groovy/io/servicetalk/gradle/plugin/internal/Versions.groovy:24]: This constant is never used, consider removing it

File: servicetalk-gradle-plugin-internal/src/main/groovy/io/servicetalk/gradle/plugin/internal/ServiceTalkLibraryPlugin.groovy [hunk 5]
```diff
@@ -206,6 +207,18 @@ final class ServiceTalkLibraryPlugin extends ServiceTalkCorePlugin {
     }
   }
 
+  private static void applyPitestPlugin(Project project) {
+    project.configure(project) {
+      pluginManager.apply("info.solidsoft.pitest")
+
+      pitest {
```
Comment (idelpivnitskiy) [inline servicetalk-gradle-plugin-internal/src/main/groovy/io/servicetalk/gradle/plugin/internal/ServiceTalkLibraryPlugin.groovy:214]: Gradle defines a group for each task. You can see them using `./gradlew tasks`.
What group this task belongs to? If it's not `verification`, consider doing something similar to pmd/spotbugs:

```
      tasks.withType(Pmd).all {
        group = "verification"
      }
```

**Review by idelpivnitskiy — COMMENTED** [review]
Didn't look at the output reports, a few comments for the plugin configuration:

File: servicetalk-gradle-plugin-internal/src/main/groovy/io/servicetalk/gradle/plugin/internal/Versions.groovy (hunk 4 above)
Comment (bondolo) [reply servicetalk-gradle-plugin-internal/src/main/groovy/io/servicetalk/gradle/plugin/internal/Versions.groovy:24]: Correct. The PITest plugin doesn't seem to support the toolVersion property.

File: servicetalk-gradle-plugin-internal/src/main/groovy/io/servicetalk/gradle/plugin/internal/ServiceTalkLibraryPlugin.groovy [hunk 6]
```diff
@@ -175,6 +176,7 @@ final class ServiceTalkLibraryPlugin extends ServiceTalkCorePlugin {
         systemProperty junit5TimeoutParamName, "$junit5Timeout"
         systemProperty "junit.jupiter.extensions.autodetection.enabled", "true"
 
+        ignoreFailures = true
```
Comment (idelpivnitskiy) [inline servicetalk-gradle-plugin-internal/src/main/groovy/io/servicetalk/gradle/plugin/internal/ServiceTalkLibraryPlugin.groovy:179]: Why is this required now?

Comment (idelpivnitskiy) [reply servicetalk-gradle-plugin-internal/src/main/groovy/io/servicetalk/gradle/plugin/internal/ServiceTalkLibraryPlugin.groovy]: Intellij IDEA has a feature of multiple changelists. Look at `Git` -> `Local Changes` tab

File: servicetalk-gradle-plugin-internal/src/main/groovy/io/servicetalk/gradle/plugin/internal/ServiceTalkLibraryPlugin.groovy [hunk 7]
```diff
@@ -209,6 +210,30 @@ final class ServiceTalkLibraryPlugin extends ServiceTalkCorePlugin {
...
+        if (project.ext.isCiBuild) {
+          outputFormats = ['XML']
+        } else {
+          outputFormats = ['HTML']
+        }
+        failWhenNoMutations = false
+      }
+
+      tasks.withType(PitestTask).all {
+        group = "verification"
```
Comment (idelpivnitskiy) [inline servicetalk-gradle-plugin-internal/src/main/groovy/io/servicetalk/gradle/plugin/internal/ServiceTalkLibraryPlugin.groovy:232]: Looks like it's already in `verification` group, we can remove this

File: servicetalk-gradle-plugin-internal/src/main/groovy/io/servicetalk/gradle/plugin/internal/ServiceTalkLibraryPlugin.groovy [hunk 8]
```diff
@@ -209,6 +210,30 @@ final class ServiceTalkLibraryPlugin extends ServiceTalkCorePlugin {
     }
   }
 
+  private static void applyPitestPlugin(Project project) {
+    project.configure(project) {
+      pluginManager.apply("info.solidsoft.pitest")
+
+      pitest {
+        pitestVersion = project.properties["pitestVersion"]
+        junit5PluginVersion = project.properties["pitestPluginJunit5Version"]
```
Comment (idelpivnitskiy) [inline servicetalk-gradle-plugin-internal/src/main/groovy/io/servicetalk/gradle/plugin/internal/ServiceTalkLibraryPlugin.groovy:219]: Oh, I like this style! Didn't know it's possible to access `gradle.properties` from here. We can remove `Versions.groovy` then

**Review by idelpivnitskiy — APPROVED** [review]
🚀

File: servicetalk-gradle-plugin-internal/src/main/groovy/io/servicetalk/gradle/plugin/internal/ServiceTalkLibraryPlugin.groovy (hunk 7 above)
Comment (bondolo) [reply servicetalk-gradle-plugin-internal/src/main/groovy/io/servicetalk/gradle/plugin/internal/ServiceTalkLibraryPlugin.groovy:232]: I will comment it out for now.

Comment (bondolo) [reply servicetalk-gradle-plugin-internal/src/main/groovy/io/servicetalk/gradle/plugin/internal/ServiceTalkLibraryPlugin.groovy]: On further investigation this snippet was only adding the task to the correct group for the `gradlew tasks` command and not configuring the task for execution. The task still needs to be invoked manually. For automatic execution it would be necessary to add it as a dependency or add it as a dependency to the quality task.
